060-run-connectors-and-tasks-in-isolated-pods.md

```diff
@@ -0,0 +1,111 @@
+
+# Run Connectors and Tasks in Isolated Pods
+
+Deploy Connectors and Tasks in isolated pods, making use of the Static Assignments feature concurrently proposed for Kafka Connect.
+
+## Current situation
+
+Currently, Strimzi supports the Connect Distributed deployment model to execute KafkaConnector workloads.
+In this deployment model, a Connect cluster is made up of workers, each consisting of a single process JVM.
+In Strimzi, these workers are given uniform resources, and the number of workers in a cluster is controlled the Deployment's replicas.
+At runtime, each worker is assigned part of the Connector & Task workload through an internal rebalance algorithm.
+This rebalance algorithm does not distinguish the resource requirements of each job, and performs unweighted scheduling across all workers.
+A single worker may be assigned multiple connectors or tasks, and these all share the resources provisioned to the hosting worker.
+
+## Motivation
+
+The Connect Distributed deployment mode makes it difficult or impossible to safely operate with a heterogeneous collection of connectors and tasks.
+It prevents external mechanisms from scheduling the resources of connectors and tasks by internally assigning jobs to workers.
+It also prevents accurately characterizing the resource usage of a single job by sharing a single JVM across multiple jobs.
+Connectors which share a single cluster have no mechanisms to avoid noisy-neighbor resource starvation or outages.
+
+Kubernetes (and Strimzi) are equipped to address the resource scheduling and isolation concerns at the process boundary.
+The upcoming Kafka Connect Static Assignments feature allows each worker in a cluster to be given a single job assignment.
+This allows an external orchestration framework (e.g. Strimzi) to control resource allocations and job scheduling.
+This functionality will be added to Apache Kafka in [KIP-987](https://cwiki.apache.org/confluence/display/KAFKA/KIP-987%3A+Connect+Static+Assignments) and released in vX.Y.
+
+## Proposal
+
+In addition to the default shared-JVM deployment mode, Strimzi should also support a isolated-JVM deployment mode.
+This mode will use the Static Assignments feature to assign connectors and tasks each to their own worker.
+This will be opt-in, and able to be performed as a live migration to an existing cluster.
+
+A KafkaConnectorSpec will accept a new optional ConnectorResourceRequirements field, to be used for the Connector instance.
```

Comment:
Would the connector resources and task resources be expected to include the JVM overhead? Or will that be something extra configured somewhere else?

Reply:
I imagined that for clarity it would forward the requirements directly to the pod with no intermediate math, so these resource requirements would all need to include the JVM overhead. Are you familiar with another situation where a k8s operator computes the resource limits in a similar way?

Reply:
It was more of a question than a suggestion we should calculate it. I'm fine if the JVM overhead is expected to be included here. Might be worth stating it in the proposal.